internal/monitor/monitor.go

```diff
@@ -0,0 +1,219 @@
+	"github.com/docker/docker/api/types/mount"
+	"github.com/docker/docker/api/types/network"
+	"github.com/docker/go-connections/nat"
+	"github.com/pkg/errors"
+	"go.uber.org/zap"
+
+	"github.com/spinup-host/spinup/config"
+	"github.com/spinup-host/spinup/internal/dockerservice"
+	"github.com/spinup-host/spinup/misc"
+)
+
+const (
+	PrometheusContainerName = "spinup-prometheus"
+	PgExporterContainerName = "spinup-postgres-exporter"
+	NetworkName             = "spinup_services"
+)
+
+// Target represents a postgres service for monitoring.
+// The assumption is that all services are running on the same host, hence only ports should differ
+type Target struct {
+	DockerNetwork string
+	ContainerName string
+	UserName      string
+	Password      string
+	Port          int
+}
+
+// BootstrapServices starts up prometheus and exporter services in docker containers
+// todo: clean up started services on SIGKILL or SIGTERM
+func (r *Runtime) BootstrapServices() error {
+	var err error
+	var promContainer *dockerservice.Container
+	var pgExporterContainer *dockerservice.Container
+	ctx := context.TODO()
+
+	defer func() {
+		if err != nil {
+			if promContainer != nil {
+				stopErr := promContainer.Stop(ctx, r.dockerClient, types.ContainerStartOptions{})
+				r.logger.Error("stopping prometheus container", zap.Error(stopErr))
+			}
+
+			if pgExporterContainer != nil {
+				stopErr := pgExporterContainer.Stop(ctx, r.dockerClient, types.ContainerStartOptions{})
+				r.logger.Error("stopping exporter container", zap.Error(stopErr))
+			}
+		}
+	}()
+
+	promContainer, err = r.dockerClient.GetContainer(ctx, PrometheusContainerName)
+	if err != nil {
+		return err
+	}
+
+	if err == nil && promContainer == nil {
+		promCfgPath, err := r.getPromConfigPath()
+		if err != nil {
+			return errors.Wrap(err, "failed to mount prometheus config")
+		}
+		promContainer, err = r.newPrometheusContainer(promCfgPath)
+		if err != nil {
+			return err
+		}
+		_, err = promContainer.Start(ctx, r.dockerClient)
+		if err != nil {
+			return errors.Wrap(err, "failed to start prometheus container")
+		}
+
+		// we expect all containers to have the same gateway IP, but we assign it here
+		// so that we can update the prometheus config with the right IP of targets
+		r.dockerHostAddr = promContainer.NetworkConfig.EndpointsConfig[NetworkName].Gateway
+		if err = r.writePromConfig(promCfgPath); err != nil {
+			return errors.Wrap(err, "failed to update prometheus config")
+		}
+	} else {
+		// if the container exists, we only update the host address without over-writing the existing prometheus config
+		r.dockerHostAddr = promContainer.NetworkConfig.EndpointsConfig[NetworkName].Gateway
+		r.logger.Info("reusing existing prometheus container")
+	}
+
+	pgExporterContainer, err = r.dockerClient.GetContainer(ctx, PgExporterContainerName)
+	if err != nil {
+		return err
+	}
+	if err == nil && pgExporterContainer == nil {
+		pgExporterContainer, err = r.newPostgresExporterContainer("")
```

Comment:
when the `dsn` is empty, then `env` is empty. How does postgresExporter knows which Postgres to connect to? (I am missing something)

Reply:
Yeah, if the dsn is empty, the exporter doesn't connect to any postgres instance and will only export go metrics (no postgres metrics). Once a new service is created, we update the dsn with the new service and recreate the container - and it can start exporting metrics for that.